### DISCUSSION TITLE
The aim of stdmicro

### DISCUSSION THREAD

**Discussion by jwkblades** [elibs/stdmicro#1] [General]
# Welcome to the discussions!

The goal of `stdmicro` is to provide C++ micro controller developers with a minimal library which can run on bare-bones hardware and provide much of the same functionality as the C++ standard template library (STL). This will also require that we have some hardware interactions, which will likely be device-specific, so keep that in mind while developing.

## Some of the core tenants:
* Resources come at a premium, keep memory allocation and data structure size to a minimum.
* Allocate from pools; this will allow developers to limit the resources (primarily RAM) available to `stdmicro`, giving them additional control over their micro controller.
* Prefer non-allocating structures when possible.
* Library functionality will be in the namespace `stdmicro`.
* Hardware functionality will be in the namespace `stdmicro::hardware`
* Unit test your code! Hardware-non-specific code should be developed in such a way that it is capable of running on any platform, and should be able to fall back to using the STL for full-fledged machines. This will allow us to verify things like proper lock usage and no race conditions in the library at higher thread counts than are likely available on the physical hardware.
* Prefer non-locking structures when possible.
* Document your code with doxygen-style comments.

## Initial purpose
While the goal is to eventually be a general (or as general as possible) purpose library, however it will start with STL-compatibility (to allow for test-ability) and the RP2040 chipset (based on ARM Cortex M0+).

The RP2040 processor datasheet can be found [here](https://datasheets.raspberrypi.org/rp2040/rp2040-datasheet.pdf).

**Comment by jwkblades** [comment]
@marshall-mcmullen @thecodenomad , I expect you two will find this library fun based on your interests. I'll be really diving in to this probably over the weekend with allocator/ memory pooling. 

I'm also planning on standing up a Makefile-based compile and output capability for the RP2040 and Linux (the Pico SDK currently uses cmake).

**Reply by marshall-mcmullen** [reply]
@jwkblades  !!!! 

I can't wait to see what you come up with and would love to participate.

I really love the idea of abstracting out lower-level hardware interactions. Sounds pretty darn interesting to me.

**Reply by jwkblades** [reply]
Agreed. With any luck, my boards should be arriving today, so I can start digging in to this a bit more tonight.

**Reply by thecodenomad** [reply]
Sweet! I'm curious to see where this goes!! :D :D 

@jwkblades what boards did you get??

**Reply by jwkblades** [reply]
I picked up the Raspberry Pi Pico (RP2040 chipset). I think I will eventually also pick up some ESP32 microcontrollers, which have Bluetooth Low Energy, and 2.4GHz WiFi capability (for about the same price as the Pico)
